sandbox_store.go

```diff
@@ -272,7 +270,7 @@ func (c *controller) sandboxCleanup(activeSandboxes map[string]interface{}) {
 				logrus.Errorf("failed to restore endpoint %s in %s for container %s due to %v", eps.Eid, eps.Nid, sb.ContainerID(), err)
 				continue
 			}
-			heap.Push(&sb.endpoints, ep)
+			sb.addEndpoint(ep)
 		}
 
```

Comment:
Unless there is a reason to not keep the sandbox locking within the methods (see above), then one *should* acquire the sandbox lock before ranging over the endpoints at the end of this function.  Alternately, one could just range over sb.getConnectedEndpoints().  Overall, it probably won't matter based on how this function gets used.

I would have put this comment closer to the actual source line, but given that it is not near a modification github won't let me.

Reply:
I generally agree, it's better to have the lock inside the functions. It didn't work for me because `clearNetworkResources` already acquired a lock when calling `sb.removeEndpoint` and go doesn't seem to like double locking. I wasn't sure if unlocking and relocking around calling `sb.removeEndpoint` is an option.

Reply:
Ah, I see.  Suggest the following then.  Still lock in removeEndpoint(), but have it wrap a method removeEndpointRaw() (rename as desired) that does the real work.  Call the latter within the sandbox methods (clearNetworkResources()) and the former outside of sandbox methods (e.g. in sbJoin, sandboxCleanup()).  So removeEndpoint() would look something like this:
```go
func (sb *sandbox) removeEndpoint(ep *endpoint) {
        sb.Lock()
        defer sb.Unlock()
        return sb.removeEndpointRaw(ep)
}
```
and with removeEndpointRaw() would look like removeEndpoint() does in your patch.  Makes sense?

Reply:
Will change the code tomorrow. As a side-effect we then also do locking when calling `sb.addEndpoint` from `sandbox_store.go`, which I assume should be fine.
Do you have any objections on #2093 b) and c)?